File: archive/j/javascript/base64-encode-decode.js
#!/usr/bin/env node

function usageAndExit() {
  console.error("Usage: please provide a mode and a string to encode/decode");
  process.exit(1);
}

function isValidBase64(str) {
  // Check length multiple of 4
  if (str.length % 4 !== 0) return false;
  // Only A–Z, a–z, 0–9, +, /, = allowed
  if (!/^[A-Za-z0-9+/]*={0,2}$/.test(str)) return false;
  // = only allowed at end, at most two
  const idx = str.indexOf("=");
  if (idx !== -1 && idx < str.length - 2) {
    // e.g. “ab==cd” invalid
    return false;
  }
  return true;
}

function encodeBase64(input) {
  // Use Buffer in Node.js
  return Buffer.from(input, "ascii").toString("base64");
}

function decodeBase64(input) {
  // Validate first
  if (!isValidBase64(input)) {
    usageAndExit();
  }
  // Buffer will ignore invalid trailing bits
  return Buffer.from(input, "base64").toString("ascii");
}

function main() {
  const args = process.argv.slice(2);
  if (args.length < 2) {
    usageAndExit();
  }
  const mode = args[0];
  const text = args[1];

  if (!text) {
    usageAndExit();
  }

  if (mode === "encode") {
    console.log(encodeBase64(text));
  } else if (mode === "decode") {
    console.log(decodeBase64(text));
  } else {
    usageAndExit();
  }
}

if (require.main === module) {
  main();
}

File: archive/j/java/Base64EncodeDecode.java
import java.nio.charset.StandardCharsets;
import java.util.Base64;

public class Base64EncodeDecode {
    public static void main(String[] args) {
        if (args == null || args.length != 2) {
            usage();
        }

        String mode = args[0];
        String input = args[1];

        if (mode == null || input == null) {
            usage();
        }

        switch (mode) {
            case "encode" -> encode(input);
            case "decode" -> decode(input);
            default -> usage();
        }
    }

    private static void encode(String input) {
        byte[] bytes = input.getBytes(StandardCharsets.UTF_8);
        String encoded = Base64.getEncoder().encodeToString(bytes);
        System.out.println(encoded);
    }


    private static void decode(String input) {
        if (!looksLikeBase64(input)) {
            usage();
        }

        try {
            byte[] decoded = Base64.getDecoder().decode(input);
            String output = new String(decoded, StandardCharsets.UTF_8);
            System.out.println(output);
        } catch (IllegalArgumentException ex) {
            usage();
        }
    }

    private static boolean looksLikeBase64(String input) {

        if (input.isEmpty() || (input.length() % 4 != 0)) {
            return false;
        }

        int len = input.length();
        int pad = 0;

        for (int i = 0; i < len; i++) {
            char c = input.charAt(i);

            if (c == '=') {
                pad++;
            } else {
                if (pad > 0) return false; // padding only at end
                if (!isBase64Char(c)) return false;
            }
        }

        int padCount = 0;
        for (int i = input.length() - 1; i >= 0 && input.charAt(i) == '='; i--) {
            padCount++;
        }

        return padCount <= 2;
    }

    private static boolean isBase64Char(char c) {
        return (c >= 'A' && c <= 'Z')
                || (c >= 'a' && c <= 'z')
                || (c >= '0' && c <= '9')
                || c == '+' || c == '/';
    }

    private static void usage() {
        System.out.println("Usage: please provide a mode and a string to encode/decode");
        System.exit(1);
    }
}